README.md

>**DEPRECATED** Please note that this repository is no longer used and will likely be removed at some point in the future. The information in this readme may be out of date or possibly misleading and should not be relied upon.

# sdx-downstream-ctp

[![Build Status](https://travis-ci.org/ONSdigital/sdx-downstream-ctp.svg?branch=develop)](https://travis-ci.org/ONSdigital/sdx-downstream-ctp) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/93d7f118c91941c9b5719c7e1d8cc9ac)](https://www.codacy.com/app/ons-sdc/sdx-downstream-ctp?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ONSdigital/sdx-downstream-ctp&amp;utm_campaign=Badge_Grade) [![codecov](https://codecov.io/gh/ONSdigital/sdx-downstream-ctp/branch/develop/graph/badge.svg)](https://codecov.io/gh/ONSdigital/sdx-downstream-ctp)

The sdx-downstream-ctp app is used within the Office for National Statistics (ONS) for consuming decrypted Survey Data Exchange (SDX) Surveys from sdx-store and delivering them to CTP.

## Installation

To install, use:

```bash
make build
```

To install using local sdx-common repo (requires SDX_HOME environment variable), use:

```bash
make dev
```

To run the test suite, use:

```bash
make test
```

It's also possible to install within a container using docker. From the sdx-downstream directory:

    $ docker build -t sdx-downstream .

## Configuration

The following envioronment variables can be set:

| Environment Variable    | Default                               | Description
|-------------------------|---------------------------------------|----------------
| SDX_STORE_URL           | `http://sdx-store:5000`               | The URL of the `sdx-store` service
| SDX_SEQUENCE_URL        | `http://sdx-sequence:5000`            | The URL of the `sdx-sequence` service
| FTP_HOST                | `pure-ftpd`                           | FTP to monitor
| FTP_USER                | _none_                                | User for FTP account if required
| FTP_PASS                | _none_                                | Password for FTP account if required
| FTP_FOLDER              | `/`                                   | FTP folder
| FTP_HEARTBEAT_FOLDER    | `/heartbeat`                          | FTP heartbeat folder
| RABBIT_QUEUE            | `sdx-ctp-survey-notifications`        | Rabbit queue name
| RABBIT_EXCHANGE         | `message`                             | RabbitMQ exchange to use
| SFTP_HOST               | _xxx.x.x.x_                           | SFTP host      
| SFTP_PORT               | `22`                                  | SFTP port

### License

Copyright © 2016, Office for National Statistics (https://www.ons.gov.uk)

Released under MIT license, see [LICENSE](LICENSE) for details.


CHANGELOG.md
### Unreleased
  - Add message to README now that this repo is deprecated

### 2.2.0 2017-07-25
  - Change all instances of ADD to COPY in Dockerfile
  - Remove use of SDX_HOME variable in makefile

### 2.1.0 2017-07-10
  - Update timestamp in all logs as UTC
  - Update to use async_consumer from sdx-common
  - Add common library logging
  - Add environment variables to README
  - Add correct license attribution
  - Add codacy badge
  - Add support for codecov to see unit test coverage 
  - Add logging binary filename before attempting delivery
  - Fix incorrect git merge in Dockerfile and requirements.txt

### 2.0.0 2017-03-15
  - Log version number on startup
  - Fix handling of None responses in remote call
  - Add SFTP support
  - Change `nack for retry` to `nack` for SDX logging
  - Change `status_code` to `status` for SDX logging
  - Change logging messages to add the service called or returned from

### 1.0.0 2017-02-16
  - Initial release


tests/test_data.py
census_survey = '''
{
  "tx_id": "0f534ffc-9442-414c-b39f-a756b4adc6cb",
  "type": "uk.gov.ons.edc.eq:surveyresponse",
  "origin": "uk.gov.ons.edc.eq",
  "version": "0.0.1",
  "survey_id": "census",
  "collection": {
    "instrument_id": "hh2016",
    "exercise_sid": "hfjdskf",
    "period": "0616"
  },
  "submitted_at": "2016-03-12T10:39:40Z",
  "metadata": {
    "user_id": "789473423",
    "ru_ref": "1234570071A"
  },
  "data": {
    "1": "2",
    "2": "4",
    "3": "2",
    "4": "Y"
  }
}
'''


tests/test_processor.py
import unittest
import mock
import json
import logging
from structlog import wrap_logger
from app.processor import CTPProcessor
from tests.test_data import census_survey
from requests import Response
from app.helpers.sdxftp import SDXFTP
from app.helpers.exceptions import BadMessageError, RetryableError

logger = wrap_logger(logging.getLogger(__name__))
ftp = SDXFTP(logger, "", "", "")


class TestCTPProcessor(unittest.TestCase):

    def setUp(self):
        survey = json.loads(census_survey)
        self.processor = CTPProcessor(logger, survey, ftp)

    def _get_response(self):
        response = Response()
        response.status_code = 200
        response._content = b'Some content'
        return response

    def test_sequence(self):
        with mock.patch('app.processor.get_sequence_no') as seq_mock:

            with mock.patch('app.helpers.sdxftp.SDXFTP.deliver_binary') as ftp_mock:
                ftp_mock.return_value = True

                response = self._get_response()
                with mock.patch('app.helpers.request_helper.remote_call') as call_mock:
                    call_mock.return_value = response

                    # Good return
                    seq_mock.return_value = '1001'
                    self.processor.process()

                    # Call issue
                    seq_mock.return_value = None
                    with self.assertRaises(RetryableError):
                        self.processor.process()

    def test_ftp(self):
        with mock.patch('app.processor.get_sequence_no') as seq_mock:
            seq_mock.return_value = '1001'

            response = self._get_response()
            with mock.patch('app.helpers.request_helper.remote_call') as call_mock:
                call_mock.return_value = response

                with mock.patch('app.helpers.sdxftp.SDXFTP.deliver_binary') as ftp_mock:

                    # Good deliver
                    self.processor.process()

                    # Failed deliver
                    ftp_mock.side_effect = RuntimeError
                    with self.assertRaises(RetryableError):
                        self.processor.process()

    def test_bad_data(self):
        with mock.patch('app.processor.get_sequence_no') as seq_mock:
            seq_mock.return_value = '1001'

            with mock.patch('app.helpers.sdxftp.SDXFTP.deliver_binary') as ftp_mock:
                ftp_mock.return_value = True

                # Empty data
                local_processor = CTPProcessor(logger, None, ftp)
                with self.assertRaises(BadMessageError):
                    local_processor.process()

    def test_name_replacer(self):
        filename = "12345.json"
        self.assertEqual(self.processor._get_completed_filename(filename), "12345.completed")


tests/test_request_helper.py
import unittest
from requests import Response
from app.helpers import request_helper


class TestRequestHelper(unittest.TestCase):

    def test_response_none(self):
        response = None
        result = request_helper.response_ok(response)
        self.assertEqual(result, False)

    def test_response_ok_200_return_true(self):
        response = Response()
        response.status_code = 200
        result = request_helper.response_ok(response)
        self.assertEqual(result, True)

    def test_response_ok_400_return_false(self):
        response = Response()
        response.status_code = 400
        result = request_helper.response_ok(response)
        self.assertEqual(result, False)

    def test_response_ok_500_return_false(self):
        response = Response()
        response.status_code = 500
        result = request_helper.response_ok(response)
        self.assertEqual(result, False)

    def test_service_name_return_responses(self):
        url = "www.testing.test/responses/12345"
        service = request_helper.service_name(url)
        self.assertEqual(service, 'SDX_STORE')

    def test_service_name_return_sequence(self):
        url = "www.testing.test/sequence"
        service = request_helper.service_name(url)
        self.assertEqual(service, 'SDX_SEQUENCE')

    def test_service_name_return_none(self):
        url = "www.testing.test/test/12345"
        service = request_helper.service_name(url)
        self.assertEqual(service, None)

    def test_service_name_url_none(self):
        url = None
        service = request_helper.service_name(url)
        self.assertEqual(service, None)


app/consumer.py
from .processor import CTPProcessor
from app import __version__
from app import settings
from app.helpers.exceptions import BadMessageError, RetryableError
from app.helpers.request_helper import get_doc_from_store
from app.helpers.sftp import SFTP
from app.settings import logger

from sdx.common.async_consumer import AsyncConsumer


def get_delivery_count_from_properties(properties):
    """
    Returns the delivery count for a message from the rabbit queue. The
    value is auto-set by rabbitmq.
    """
    delivery_count = 0
    if properties.headers and 'x-delivery-count' in properties.headers:
        delivery_count = properties.headers['x-delivery-count']
    return delivery_count + 1


class Consumer(AsyncConsumer):

    def __init__(self):
        self._ftp = SFTP(
            logger, settings.SFTP_HOST, settings.SFTP_USER,
            port=int(settings.SFTP_PORT), privKey=settings.SFTP_PRIVATEKEY_FILENAME
        )
        super(Consumer, self).__init__()

    def on_message(self, unused_channel, basic_deliver, properties, body):

        try:
            tx_id = body.decode("utf-8")

            logger.info("Decoded message body",
                        delivery_tag=basic_deliver.delivery_tag,
                        tx_id=tx_id)
        except UnicodeDecodeError as e:
            logger.error("Cannot decode message body",
                         queue=self.QUEUE,
                         delivery_tag=basic_deliver.delivery_tag,
                         app_id=properties.app_id,
                         error=e)

            self.reject_message(basic_deliver.delivery_tag)
            return None

        delivery_count = get_delivery_count_from_properties(properties)

        logger.info(
            'Received message',
            queue=self.QUEUE,
            delivery_tag=basic_deliver.delivery_tag,
            delivery_count=delivery_count,
            app_id=properties.app_id
        )

        document = get_doc_from_store(tx_id)
        processor = CTPProcessor(logger, document, self._ftp)

        try:
            processor.process()
            self.acknowledge_message(basic_deliver.delivery_tag, tx_id=processor.tx_id)
            logger.info("Processed successfully", tx_id=processor.tx_id)

        except BadMessageError as e:
            # If it's a bad message then we have to reject it
            self.reject_message(basic_deliver.delivery_tag, tx_id=processor.tx_id)
            logger.error("Bad message", action="rejected", exception=e, delivery_count=delivery_count, tx_id=processor.tx_id)

        except (RetryableError, Exception) as e:
            self.nack_message(basic_deliver.delivery_tag, tx_id=processor.tx_id)
            logger.error("Failed to process", action="nack", exception=e, delivery_count=delivery_count, tx_id=processor.tx_id)


def main():
    logger.info("Starting consumer", version=__version__)
    consumer = Consumer()
    try:
        consumer.run()
    except KeyboardInterrupt:
        consumer.stop()

if __name__ == '__main__':
    main()


app/processor.py
from app.helpers.request_helper import get_sequence_no
import json
from app import settings
import re
from app.helpers.exceptions import BadMessageError, RetryableError


class CTPProcessor(object):

    def __init__(self, logger, survey, ftpconn):
        self.logger = logger
        self.survey = survey
        self.tx_id = None
        self._setup_logger()
        self.ftp = ftpconn

    def process(self):
        sequence_no = get_sequence_no()
        if sequence_no is None:
            raise RetryableError("Failed to get sequence number")

        filename = '{}.json'.format(sequence_no)

        if self.survey is None:
            raise BadMessageError("No survey data")

        data = json.dumps(self.survey)

        # Attempt to deliver the real file and send a .completed after it
        self._deliver_file(filename, data)
        self._deliver_file(self._get_completed_filename(filename), "")

        return

    def _setup_logger(self):
        if self.survey:
            if 'metadata' in self.survey:
                metadata = self.survey['metadata']
                self.logger = self.logger.bind(user_id=metadata['user_id'], ru_ref=metadata['ru_ref'])

            if 'tx_id' in self.survey:
                self.tx_id = self.survey['tx_id']
                self.logger = self.logger.bind(tx_id=self.tx_id)

    def _deliver_file(self, filename, data):
        self.logger.debug("Attempting to deliver to ftp", filename=filename)
        folder = self._get_ftp_folder(self.survey)

        try:
            self.ftp.deliver_binary(folder, filename, data.encode('utf-8'))
        except (RuntimeError, Exception) as e:
            self.logger.error("Failed to deliver file to ftp", filename=filename, exception=e)
            raise RetryableError("Failed to deliver file to ftp")

        self.logger.info("Delivered file to ftp", filename=filename)

    def _get_completed_filename(self, filename):
        return re.compile('.json$').sub('.completed', filename)

    def _get_ftp_folder(self, survey):
        if 'heartbeat' in survey and survey['heartbeat'] is True:
            return settings.FTP_HEARTBEAT_FOLDER
        else:
            return settings.FTP_FOLDER


app/helpers/sftp.py
import io
import os.path
import tempfile
import subprocess
import sys
import zipfile

__doc__ = """
Example:

python -m app.helpers.sftp < test-output.zip

"""


class SFTP:

    @staticmethod
    def operations(locn, home=".", mkdirs=False):
        """
        Returns a sequence of commands to copy a file tree from local 'locn' to server 'home'.

        """
        rv = []
        pos = len(locn.split(os.sep))
        for n, (root, dirs, files) in enumerate(os.walk(locn)):
            # We are walking the local file tree
            if n == 0:
                rv.append("cd {0}\n".format(home))
                crumbs = []
            else:
                crumbs = [".."] * (len(root.split(os.sep)) - pos)

            # Navigate to the same spot remotely
            dest = root.split(os.sep)[pos:]
            if dest:
                rv.append("cd {0}\n".format(os.path.join(*dest)))

            if mkdirs:
                rv.extend(["mkdir {0}\n".format(os.path.basename(i)) for i in dirs])

            # Place files
            rv.extend(["put {0}\n".format(os.path.join(root, i)) for i in files])

            # Go back to the home location
            if crumbs:
                rv.append("cd {0}\n".format(os.path.join(*crumbs)))

        return rv

    @staticmethod
    def transfer(cmds, user, host, port, priv_key=None, quiet=True):
        """
        Connects to an sftp server and plays a sequence of commands.

        """
        args = [
            "sftp", "-o", "ControlPersist=true", "-o", "ControlMaster=auto",
            "-o", "ControlPath=~/.ssh/ssh-%r@%h:%p", "-b", "-", "-P", str(port),
            "{user}@{host}".format(user=user, host=host)
        ]
        if priv_key is not None:
            args[-2:-1] = [str(port), "-i", priv_key]
        kwargs = {"stdout": subprocess.DEVNULL, "stderr": subprocess.DEVNULL} if quiet else {}
        with subprocess.Popen(args, stdin=subprocess.PIPE, **kwargs) as proc:
            for cmd in cmds:
                proc.stdin.write(cmd.encode("utf-8"))
            proc.stdin.write("bye\n".encode("utf-8"))
        return proc.returncode

    def __init__(self, logger, host, user, priv_key=None, port=22):
        self.logger = logger
        self.host = host
        self.user = user
        self.priv_key = priv_key
        self.port = port

    def deliver_binary(self, folder, filename, data):
        self.logger.info("Delivering binary file to FTP", host=self.host, folder=folder, filename=filename)
        with tempfile.NamedTemporaryFile(dir="tmp") as locn:
            locn.write(data)
            locn.flush()
            cmds = [
                "cd {0}\n".format(folder),
                "put {0} {1}\n".format(locn.name, filename),
                "bye\n"
            ]
            rv = self.transfer(
                cmds, user=self.user, host=self.host, port=self.port,
                priv_key=self.priv_key, quiet=True
            )
        if rv != 0:
            msg = "Failed to deliver file to FTP"
            self.logger.error(msg, host=self.host)
            raise UserWarning(msg)

    def unzip_and_deliver(self, folder, zip_contents):
        with tempfile.TemporaryDirectory(dir="tmp") as locn:
            with zipfile.ZipFile(io.BytesIO(zip_contents)) as payload:
                payload.extractall(locn)
                cmds = self.operations(locn)
                rv = self.transfer(
                    cmds, user=self.user, host=self.host, port=self.port,
                    priv_key=self.priv_key, quiet=True
                )
        if rv != 0:
            msg = "Failed to deliver file to FTP"
            self.logger.error(msg, host=self.host)
            raise UserWarning(msg)


def run():
    """
    Reads a ZIP file or text file from stdin and copies the contents to the correct
    location on an SFTP server.

    Parameters are configured to match the test container in sdx-compose.

    """
    subprocess.check_output(
        ["ssh-keygen", "-f", os.path.expanduser("~/.ssh/known_hosts"), "-R", "[0.0.0.0]:2222"]
    )

    user = "testuser"
    host = "0.0.0.0"
    port = 2222
    root = "public"

    try:
        with zipfile.ZipFile(sys.stdin.buffer) as payload:
            with tempfile.TemporaryDirectory() as locn:
                payload.extractall(locn)
                cmds = SFTP.operations(locn, home="public", mkdirs=True)
                rv = SFTP.transfer(
                    cmds, user="testuser", host="0.0.0.0", port=2222, quiet=False
                )
    except zipfile.BadZipFile:
        # Assume a text file
        sys.stdin.buffer.seek(0)
        data = sys.stdin.buffer.read()
        with tempfile.NamedTemporaryFile() as locn:
            locn.write(data)
            locn.flush()
            cmds = [
                "cd {0}\n".format(root),
                "put {0} {1}\n".format(locn.name, "data.txt"),
                "bye\n"
            ]
            rv = SFTP.transfer(cmds, user, host, port, quiet=False)
    return rv


if __name__ == "__main__":
    rv = run()
    sys.exit(rv)


app/helpers/ftp_helper.py
from app import settings
import io
from ftplib import FTP


def get_ftp_folder(survey):
    if 'heartbeat' in survey and survey['heartbeat'] is True:
        return settings.FTP_HEARTBEAT_FOLDER
    else:
        return settings.FTP_FOLDER


def connect_to_ftp():
    ftp = FTP(settings.FTP_HOST)
    ftp.login(user=settings.FTP_USER, passwd=settings.FTP_PASS)
    return ftp


def process_file_to_ftp(folder, filename, data):
    try:
        settings.logger.debug("Processing file", folder=folder, filename=filename)
        ftp = connect_to_ftp()
        deliver_binary_to_ftp(ftp, folder, filename, bytes(data, 'UTF-8'))
        ftp.quit()
        return True

    except (RuntimeError) as e:
        settings.logger.error("Exception processing file", exception=e)
        return False


def deliver_binary_to_ftp(ftp, folder, filename, data):
    stream = io.BytesIO(data)
    ftp.cwd(folder)
    ftp.storbinary('STOR ' + filename, stream)


app/helpers/request_helper.py
from app.settings import logger, session, SDX_SEQUENCE_URL, SDX_STORE_URL
from requests.packages.urllib3.exceptions import MaxRetryError
from requests.exceptions import ConnectionError
from app.helpers.exceptions import RetryableError


def service_name(url=None):
    try:
        parts = url.split('/')
        if 'responses' in parts:
            return 'SDX_STORE'
        elif 'sequence' in parts:
            return 'SDX_SEQUENCE'
    except AttributeError as e:
        logger.error(e)


def remote_call(url, json=None):
    service = service_name(url)

    try:
        logger.info("Calling service", request_url=url, service=service)
        response = None

        if json:
            response = session.post(url, json=json)
        else:
            response = session.get(url)

        return response

    except MaxRetryError:
        logger.error("Max retries exceeded (5)", request_url=url)
        raise RetryableError("Max retries exceeded")
    except ConnectionError:
        logger.error("Connection error", request_url=url)
        raise RetryableError("Connection error")


def response_ok(response, service_url=None):
    service = service_name(service_url)

    if response is None:
        logger.error("No response from service")
        return False
    elif response.status_code == 200:
        logger.info("Returned from service", request_url=response.url, status=response.status_code, service=service)
        return True
    else:
        logger.error("Returned from service", request_url=response.url, status=response.status_code, service=service)
        return False


def get_sequence_no():
    sequence_url = "{0}/json-sequence".format(SDX_SEQUENCE_URL)
    response = remote_call(sequence_url)
    if not response_ok(response, sequence_url):
        return None

    result = response.json()
    return result.get('sequence_no')


def get_doc_from_store(tx_id):
    store_url = "{0}/responses/{1}".format(SDX_STORE_URL, tx_id)
    response = remote_call(store_url)

    if not response_ok(response, store_url):
        return None

    return response.json()


app/processors/common_software_processor.py
from app import settings
from app.helpers.request_helper import remote_call, response_ok, get_sequence_no
from app.helpers.ftp_helper import get_ftp_folder, process_file_to_ftp


class CTPProcessor(object):

    def __init__(self, logger, survey):
        self.logger = logger
        self.survey = survey
        self.tx_id = None
        self.setup_logger()

    def setup_logger(self):
        if self.survey:
            if 'metadata' in self.survey:
                metadata = self.survey['metadata']
                self.logger = self.logger.bind(user_id=metadata['user_id'], ru_ref=metadata['ru_ref'])

            if 'tx_id' in self.survey:
                self.tx_id = self.survey['tx_id']
                self.logger = self.logger.bind(tx_id=self.tx_id)

    def get_url(self):
        sequence_no = get_sequence_no()
        return "{0}/census/{1}".format(settings.SDX_TRANSFORM_CTP_URL, sequence_no)

    def transform(self):
        response = remote_call(self.get_url(), json=self.survey)
        if not response or not response_ok(response):
            return None

        return response.content

    def deliver_file(self, data):
        folder = get_ftp_folder(self.survey)
        return process_file_to_ftp(folder, data)

    def process(self):
        transformed = self.transform()

        if transformed is None:
            return False

        return self.deliver_file(transformed)


app/settings.py
import logging
from structlog import wrap_logger
import os.path
import pwd
import requests
from requests.packages.urllib3.util.retry import Retry
from requests.adapters import HTTPAdapter

from sdx.common.logger_config import logger_initial_config

LOGGING_LEVEL = logging.getLevelName(os.getenv('LOGGING_LEVEL', 'DEBUG'))

logger_initial_config(service_name='sdx-downstream-ctp',
                      log_level=LOGGING_LEVEL)

logger = wrap_logger(logging.getLogger(__name__))

APP_ROOT = os.path.dirname(os.path.abspath(__file__))
APP_TMP = os.path.join(APP_ROOT, 'tmp')

# Default to true, cast to boolean
SDX_STORE_URL = os.getenv("SDX_STORE_URL", "http://sdx-store:5000")
SDX_SEQUENCE_URL = os.getenv("SDX_SEQUENCE_URL", "http://sdx-sequence:5000")

FTP_HOST = os.getenv('CTP_FTP_HOST', 'pure-ftpd')
FTP_USER = os.getenv('CTP_FTP_USER')
FTP_PASS = os.getenv('CTP_FTP_PASS')

FTP_FOLDER = os.getenv('CTP_FTP_FOLDER', '/')
FTP_HEARTBEAT_FOLDER = os.getenv('CTP_FTP_HEARTBEAT_FOLDER', '/heartbeat')

SFTP_HOST = os.getenv("SFTP_HOST", "127.0.0.1")
SFTP_PORT = os.getenv("SFTP_PORT", "22")
SFTP_USER = os.getenv('SFTP_USER', pwd.getpwuid(os.getuid())[0])
SFTP_PRIVATEKEY_FILENAME = os.getenv(
    "SFTP_PRIVATEKEY_FILENAME", os.path.expanduser(os.path.join("~", ".ssh", "id_rsa"))
)
SFTP_PUBLICKEY_FILENAME = os.getenv(
    "SFTP_PUBLICKEY_FILENAME", os.path.expanduser(os.path.join("~", ".ssh", "id_rsa.pub"))
)

RABBIT_QUEUE = os.getenv('CTP_NOTIFICATIONS_QUEUE', 'sdx-ctp-survey-notifications')
RABBIT_EXCHANGE = os.getenv('RABBITMQ_EXCHANGE', 'message')

RABBIT_URL = 'amqp://{user}:{password}@{hostname}:{port}/{vhost}'.format(
    hostname=os.getenv('RABBITMQ_HOST', 'rabbit'),
    port=os.getenv('RABBITMQ_PORT', 5672),
    user=os.getenv('RABBITMQ_DEFAULT_USER', 'rabbit'),
    password=os.getenv('RABBITMQ_DEFAULT_PASS', 'rabbit'),
    vhost=os.getenv('RABBITMQ_DEFAULT_VHOST', '%2f')
)

RABBIT_URL2 = 'amqp://{user}:{password}@{hostname}:{port}/{vhost}'.format(
    hostname=os.getenv('RABBITMQ_HOST2', 'rabbit'),
    port=os.getenv('RABBITMQ_PORT2', 5672),
    user=os.getenv('RABBITMQ_DEFAULT_USER', 'rabbit'),
    password=os.getenv('RABBITMQ_DEFAULT_PASS', 'rabbit'),
    vhost=os.getenv('RABBITMQ_DEFAULT_VHOST', '%2f')
)

RABBIT_URLS = [RABBIT_URL, RABBIT_URL2]

# Configure the number of retries attempted before failing call
session = requests.Session()

retries = Retry(total=5, backoff_factor=0.1)

session.mount('http://', HTTPAdapter(max_retries=retries))
session.mount('https://', HTTPAdapter(max_retries=retries))
